### ISSUE TITLE
Discuss: Extend ∧ to concatenate lists

### ISSUE THREAD

**Issue by Gabriel439** [dhall-lang/dhall-lang#754]
This idea is pretty simple: change `∧` to permit concatenating two list-valued fields of the same type.  For example:

```dhall
{ foo = [ 1 ], bar = True } ∧ { foo = [ 2 ] }

= { foo = [ 1, 2 ], bar = True }
```

In fact, this would actually make the `∧` operator supersede the `#` operator, because this would now also be valid:

```
[ 1 ] ∧ [ 2 ]

= [ 1, 2 ]
```

~~There is also a theoretically elegant reason for doing this, which is that it makes `toMap` a monoid homomorphism:~~

```dhall
toMap (x ∧ y) = toMap x ∧ toMap y

toMap {} : List { mapKey Text : mapValue : T } = [] : List { mapKey Text : mapValue : T }
```

The practical motivation for this is to better support `dhall-kubernetes`.  A common idiom in the Helm world is add an `extraLabels` configuration option to inject additional labels, but if `∧` supported concatenation, then you trivially extend an existing Kubernetes resource with extra labels by just merging them in using `∧`.

For example, if you wanted to extend a starting Dhall "chart" to add extra labels to an existing `kubernetes.Deployment` you would do:

```dhall
  someDeployment
∧ { metadata =
      { labels =
          toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
      }
  } 
```

**Comment by quasicomputational** [comment]
I'm -0 on this.

This is opinionated behaviour that might not do what you want, and which is hard to opt out of. Sometimes list concatenation is what you want - but other times you do want a different monoid, like Last or First. And the desired behaviour might vary from one field to another!

Further, comparing it to present behaviour of `/\`, where the behaviour (where defined) is entirely deriveable from parametricity and the knowledge that it doesn't know anything about anything except records, it seems rather like a layering violation. Being able to peer inside value types really is quite a fundamental change.

Or, somewhat differently: `/\` as-is cannot be written in user code. It has to be a primitive; user code cannot abstract over record types. List concatenation can be written as user code (with less syntax sugar).

None of that's the most convincing, but something about this feels wrong. Maybe it'd be better to have a way to specify how to merge fields present in both sides, which could have even more general application.

**Comment by sjakobi** [comment]
> There is also a theoretically elegant reason for doing this, which is that it makes toMap a monoid homomorphism:
>
>     toMap (x ∧ y) = toMap x ∧ toMap y

I might be misunderstanding the proposal, but it doesn't look like this property would hold:

    toMap ({ foo = [1] } /\ { foo = [2] })
    = [ { mapKey = "foo", mapValue = [1, 2] } ]

    toMap { foo = [1] } /\ toMap { foo = [2] }
    = [ { mapKey = "foo", mapValue = [1] }, { mapKey = "foo", mapValue = [2] } ]


> For example, if you wanted to extend a starting Dhall "chart" to add extra labels to an existing `kubernetes.Deployment` you would do:
> 
> ```dhall
>   someDeployment
> ∧ { metadata =
>       { labels =
>           toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
>       }
>   } 
> ```

What does this look like with the current feature set?

A very small disadvantage of the proposal is that the analogy between `/\` and `//\\` is weakened.

How about, instead of generalising `/\`, we generalise `#` to also append lists in records?

```
{ foo = [ 1 ], bar = True } # { foo = [ 2 ] }

= { foo = [ 1, 2 ], bar = True }
```

I don't think `#` should be allowed to insert new fields into a record. You'd still have to use `/\` for that.

Or maybe we should add a new operator…

**Comment by Gabriel439** [comment]
@quasicomputational: What about @sjakobi's idea about overloading `#` instead of `∧`.  By using `#` the intent is clear that the user wants to resolve conflicts via list concatenation rather than some other `Monoid`.

In fact, you could imagine overloading other operators in this way, too (e.g. `+`/`*` can merge records recursively, resolving conflicts on `Natural` numbers).  Also, the user would be able to specify which behavior they want on a field-by-field basis by using different operators in succession.

For example:

```dhall
(baseConfig + { foo = 2 }) * { bar = 3 }
```

... could add `2` to the `foo` field and multiply the `bar` field by `3`.

@sjakobi: Also, you're correct that the operator would not satisfy the monoid morphism laws

**Comment by sjakobi** [comment]
> In fact, you could imagine overloading other operators in this way, too (e.g. `+`/`*` can merge records recursively, resolving conflicts on `Natural` numbers).

"Lifting" operations to operate on record fields seems pretty neat. It seems like that would at least partially cover the use cases for the lens/traversal ideas from https://github.com/dhall-lang/dhall-lang/issues/340#issuecomment-453370637.

I'm not so convinced that overloading the existing operators is the best way to achieve that though. IMHO overloading also makes it harder to understand what an operator expression means: Is `a * b` a record or a `Natural`?

The alternative would be to add explicit syntax for lifted operators, e.g. `{#}`, `{*}`, `{//}` etc.

**Comment by Gabriel439** [comment]
@sjakobi: Yeah, I'd be fine with adding new operators that are lifted equivalents instead of overloading the existing ones.  It also makes it easier to standardize their behavior because we can just standardize the behavior of lifting itself instead of standardizing how to lift each operator individually.

**Comment by sjakobi** [comment]
> The alternative would be to add explicit syntax for lifted operators, e.g. `{#}`, `{*}`, `{//}` etc.

In analogy, there could be `<#>`, `<*>`, `<//>` etc. for unions:

    let U = < foo : Bool >

    in U.foo False <||> U.foo True

=>

    U.foo True

**Comment by Gabriel439** [comment]
@sjakobi: Yeah, I like the `{_}` syntax for lifting operators over records since (and `<_>` for unions, too), although I'll probably only try to standardize `{_}` at first

**Comment by philandstuff** [comment]
We could also have `[+]` as an alias for `List/zipWith +` (not sure how serious I am with this suggestion 😉)

**Comment by Nadrieril** [comment]
The idea looks neat but doesn't seem to compose well. What if I have a record with both lists and naturals, and I want to concatenate the lists and add the naturals ?
I feel like what we really want is a generic way to operate on records/record types. Now that we have dependent types, could we imagine a way to manipulate a record or a record type using dhall functions, so that `//\\`, `/\`, `//` and `toMap` could be defined by users ? Maybe `Type/fold` and `Type/build` functions ? Or at least `RecordType/fold` and `RecordType/build`. That would kill parametricity if we do it naively, but that can be worked around.

**Comment by Gabriel439** [comment]
@Nadrieril: You would use separate operators for the lists versus the naturals, like this:

```dhall
(baseRecord {#} { foo = [ 1 ] }) {+} { bar = 2 }
```

... which would extend the `baseRecord` by concatenating `[ 1 ]` to the field `foo` and adding `2` to the field `bar`.

**Comment by singpolyma** [comment]
The initial proposal here seems reasonable to me, since currently it's just a type error so it doesn't break anything and could be useful.

The proposal this seems to be moving towards (adding a bunch of new operators as special cases to notionally "lift" operations) seems both confusing and complex.

**Comment by Gabriel439** [comment]
@singpolyma: The idea isn't so much to add N new operators but rather to add a new `{op}` construct for lifting an operator `op`.  The idea is that you can define the type-checking and normalization rules for `{op}` in terms of the type-checking and normalization rules for `op`.

**Comment by Nadrieril** [comment]
Do we want that to work for arbitrary user functions as well ? Then it would just be notation for a sort of `zipWith` for records, that's be cool

**Comment by sjakobi** [comment]
> Do we want that to work for arbitrary user functions as well ?

That would be neat, but I don't think that the `{_}` syntax would be a good fit:

    { foo = "bla" } {\(t : Text) -> \(n : Natural) -> t ++ Natural/show n} { foo = 3 }
    = { foo = "bla3" }

At least it would make parsing more tricky.

I'd rather have a keyword `lift{R,U,L}`(?) so I could write

    liftR (\(t : Text) -> \(n : Natural) -> t ++ Natural/show n) { foo = "bla" } { foo = 3 }

Maybe this could also be used with operators, e.g.

    liftR (+) { foo = 1 } { foo = 2 }
    = { foo = 3 }

**Comment by sjakobi** [comment]
Looking back at the original issue description, the main motivation seems to be the modification of fields in record.

It seems to me that a variant of `merge` might fit the bill:

```dhall
modify
  { a = λ(n : Natural) → n * 2
  , b = { c = λ(ts : [ Text ]) → ts # [ "bar" ] }
  }
  { a = 21
  , b = { c = [ "foo" ] }
  , d = True
  }

=>

{ a = 42
, b = { c = ["foo", "bar"] }
, d = True
}
```

This way we're not limited to lifting one operator at a time, and can also use custom functions.

**Comment by sjakobi** [comment]
Actually the output type need not be the same as the input type. This could be legal too:

```dhall
modify
  { a = λ(n : Natural) → { foo = n } }
  { a = 1 }

=>

{ a = { foo = 1 } }

**Comment by sjakobi** [comment]
The motivating example would be written as

```dhall
modify
  { metadata =
      { labels =
            λ(xs : Map.Entry Text Text)
          → xs # toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
      }
  }
  someDeployment
```

**Comment by Gabriel439** [comment]
@sjakobi: The main issue I can see is that the `modify` keyword has no effect if the target field is missing from the record being modified (e.g. if `someDeployment` has no `metadata.labels` field).

You could try to work around it by extending the example to:

```dhall
modify
  { metadata =
      { labels =
            λ(xs : Map.Type Text Text)
          → xs # toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
      }
  }
  ({ metadata = { labels = [] : Map.Type Text Text } } ∧ someDeployment)
```

... but then you have the opposite problem which is that the example breaks if `someDeployment` goes back to having a `labels` field.

**Comment by sjakobi** [comment]
I see. I had actually imagined `modify` not to typecheck in the case of unused handlers, but I understand that it's useful

How about a derivative of [`Data.Map.alter`](http://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Map-Lazy.html#v:alter) then?

```haskell
alter :: Ord k => (Maybe a -> Maybe a) -> k -> Map k a -> Map k a
```

```dhall
let List/default =
        λ(t : Type)
      → λ(o : Optional (List t))
      → Optional/fold (List t) o (List t) (λ(l : List t) → l) ([] : List t)

let Labels = https://prelude.dhall-lang.org/Map/Type Text Text

in  alter
      { metadata =
          { labels =
                λ(optLabels : Optional Labels)
              →   List/default Labels optLabels
                # toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
          }
      }
      someDeployment
```

Handlers would have type `Optional a -> b`. I believe we can't allow deletions (via `Optional b`) because the output type would depend on the field values then.

I guess this doesn't look too ergonomic yet, but it should be pretty flexible. We could probably add helpers for common handler patterns to the Prelude.

**Comment by Gabriel439** [comment]
@sjakobi: I think it needs to be ergonomic, though, but it doesn't need to be general.  We already have a general solution, which is to do nested overrides like this:

```dhall
  someDeployment
⫽ { metadata =
        someDeployment.metadata
      ⫽ { labels =
                someDeployment.metadata.labels
            ++  toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
        }
  }
```

However, thinking about this more, the conclusion I came to was that the number of common extension points in Kubernetes is small enough that I might be able to add utilities for this purpose (e.g. create an `addLabels` function) instead of adding new operators.

**Comment by sjakobi** [comment]
> ```haskell
>   someDeployment
> ⫽ { metadata =
>         someDeployment.metadata
>       ⫽ { labels =
>                 someDeployment.metadata.labels
>             ++  toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
>         }
>   }
> ```

This form requires that `someDeployment` contains a `metadata` field which contains a `labels` field. The `alter`-version could also work on an empty record:

```haskell
alter
  { metadata =
      { labels =
            λ(optLabels : Optional Labels)
          →   List/default Labels optLabels
            # toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
      }
  }
  {=}

=>

{ metadata = { labels = [ { mapKey = "scheduler.alpha.kubernetes.io/critical-pod", mapValue = "" } ] } }

**Comment by alexhumphreys** [comment]
> However, thinking about this more, the conclusion I came to was that the number of common extension points in Kubernetes is small enough that I might be able to add utilities for this purpose (e.g. create an `addLabels` function) instead of adding new operators.

I think one of the things that burned me last time I tried to use dhall for k8s was a deployment that had one port open on a production env (http endpoint port) and two ports on developer envs (http endpoint port and an admin endpoint port). Since ports are nested (in `jq` syntax) like so in deployments `spec.template.spec.containers | .[] | .ports`, and in services they are also an array of ports, it would have taken a lot of work to configure this in dhall. It would have been easier to use dhall to create the deployment with one port, then "hit it with a hammer" afterwards so to speak and add the second port with some kind of `∧`/`alter`/`modify` operator.

So, while I'm not sure I understand that `addLabels` suggestion, k8s configuration is labyrinthine enough that I think a more general solution is worth considering from the start.

Maybe most people do just add labels or do other relatively straight forward amendments to k8s yams files, so a few utilities would solve most of the issues. But if I am to migrate [this](https://github.com/advancedtelematic/ota-community-edition/tree/master/templates/services) thousand or so lines of yaml to dhall, I'd need to be sure I haven't backed myself into a corner, and that any later amendments to the config files are possible.

**Comment by Profpatsch** [comment]
> The initial proposal here seems reasonable to me, since currently it's just a type error so it doesn't break anything and could be useful.

That thinking is dangerous, because type errors are useful information that you are not doing what you think you are doing. If it suddenly works, but is not doing what you are doing, then you have a problem.

I’m -1 on overloading operators, because it complicates their semantics and makes it hard to reason about edge cases.

This specific case doesn’t sound sensible, because
1. as others noted, there’s multiple monoids for []
2. it’s specific to lists

> The idea isn't so much to add N new operators but rather to add a new `{op}` construct for lifting an operator `op`. The idea is that you can define the type-checking and normalization rules for `{op}` in terms of the type-checking and normalization rules for `op`.

That sounds interesting, but only if we can find a non-arbitrar way to define what “lifting” means for all operators which implement lifting. The lifting should also follow the same rules for each operator.

**Comment by Gabriel439** [comment]
This is why I prefer to wait a bit before progressing more on this.

Just to provide more context, I've been working on porting Helm's `charts` repository, both for its own sake (to promote Dhall's use for Kubernetes) and also to stress test the language to inform the language design.  I'm going to port more charts to help inform my thoughts on this proposal a bit more.

**Comment by alexhumphreys** [comment]
Oh cool, I was thinking recently dhall would need an equivalent of that repo should it want to make a dent in the k8s landscape. Guess it's for learning at the moment, but in the future if you ever need extra people working on porting charts hit me up.

**Comment by singpolyma** [comment]
> That thinking is dangerous, because type errors are useful information, in that they tell you that you are not doing what you think you are doing. If something suddenly works, but that something is not doing what you think it should be doing, then you have a problem.

This is very close to be an argument against all polymorphism, isn't it?  If something "works, but that something is not doing what you think it should be doing" that's a failure of documentation, not of the type system.

**Comment by marcosh** [comment]
Today I found myself needing something similar to this (see https://discourse.dhall-lang.org/t/record-merge-with-list-concatenation/127).

From my point of view, the optimal thing would be a to have a generic function like

```
merge
  :  (Bool       -> Bool       -> Bool      )
  -> (Natural    -> Natural    -> Natural   )
  -> (Integer    -> Integer    -> Integer   )
  -> (Double     -> Double     -> Double    )
  -> (Text       -> Text       -> Text      )
  -> (List a     -> List a     -> List a    )
  -> (Optional b -> Optional b -> Optional b)
  ->  Record     -> Record     -> Record
```

which would use the provided functions to merge possibly duplicated fields (and recur if the duplicated field contains a record).

**Comment by sjakobi** [comment]
@marcosh A limitation of your `merge` would be that e.g. any `Bool` fields would be handled the same – you couldn't select a different combining function for different fields.

How about something like this instead:

`fcombine` (just a temporary name) would take a record of binary "handlers", similarly to `merge`, and two records:

```dhall
fcombine
  { foo = \(x : Bool) -> \(y : Bool) -> x && y
  , bar = \(x : Bool) -> \(y : Bool) -> x || y
  }
  { foo = True, bar = True }
  { foo = False, bar = False }

=>

{ foo = False, bar = True }
```

If a field that we have a handler for appears in only one of the records, the handler is unused:

```dhall
fcombine
  { foo = \(x : Bool) -> \(y : Bool) -> x && y
  , bar = \(x : Bool) -> \(y : Bool) -> x || y
  }
  { foo = True, bar = True }
  { bar = False }

=>

{ foo = True, bar = True }
```

I believe that would address @Gabriel439's [criticism](https://github.com/dhall-lang/dhall-lang/issues/754#issuecomment-537310772) of my `modify` idea above.

If the records have the same field, but no handler for it, that's a type error:

```dhall
fcombine
  { foo = \(x : Bool) -> \(y : Bool) -> x && y }
  { foo = True, bar = True }
  { foo = True, bar = False }

=>

Type error: No handler for "bar"
```

I think we'd also want a type error if the handler type doesn't match:

```dhall
fcombine
  { foo = \(x : Bool) -> \(y : Bool) -> x && y }
  { foo = True }
  { foo = 1 }

=>

Type error: Type mismatch in the handler for foo
```


`fcombine` _could_ handle fields of different types:

```dhall
fcombine
  { foo = \(x : Integer) -> \(y : Natural) -> Integer/clamp x + y }
  { foo = +2 }
  { foo = 1 }

=>

{ foo = 3 }
```

… although that would allow more variety in the output record types.

If the record of handlers is `{=}`, `fcombine` could work just like `/\`:

```dhall
fcombine
  {=}
  { foo = True, baz = { _1 = "x" } }
  { bar = False, baz = { _2 = "y" } }

=>

{ foo = True, bar = False, baz = { _1 = "x", _2 = "y" } }
```

**Comment by sjakobi** [comment]
With `fcombine`, the original example would look like this:

```dhall
fcombine
  { metadata = { labels = \(x : Map Text Text) -> \(y : Map Text Text) -> x # y } }
  someDeployment
  { metadata =
      { labels =
          toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
      }
  }
```

This isn't very concise, but it's arguably simpler than using nested overrides:

```dhall
  someDeployment
⫽ { metadata =
        someDeployment.metadata
      ⫽ { labels =
              someDeployment.metadata.labels
            # toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
        }
  }
```

**Comment by matheus23** [comment]
`fcombine` feels a lot like haskell's `liftA2` to me. (Imagining something of a type of `Record (a -> b -> c) -> Record a -> Record b -> Record c`).

Soo.. Borrowing from that, alternative ways to implement said features could be something like `recordApply : Record (a -> b) -> Record a -> Record b`. And maybe also have something like `recordPure : a -> Record a`?

I have no idea whether what I wrote even makes sense. I'm just throwing vague Ideas out here, I like to provide some additional perspectives. So this is ment to be inspirational, not actual suggestions :)

**Comment by Gabriel439** [comment]
I think proposals should illustrate how they would handle the following situation based on the example by @alexhumphreys.  Suppose I had an `example` value of the following type:

```dhall
example : { spec : { template : { spec : { containers : List { ports : List Natural } } } } }
```

How would I add port `22` to each container?

**Comment by Profpatsch** [comment]
All of that just sounds like an unprincipled specialized version of recursion schemes to me.

**Comment by Gabriel439** [comment]
So if the original requirement is to make deep nested updates, another approach is to add something like lenses to the language.  In other words, if we wanted to port 22 to each container, we would write something like this:

```dhall
let Container = { ports : List Natural }

in  over
      spec.template.spec.containers
      (   λ(xs : List Container)
        → Prelude.List.map
            Container
            Container
            (   λ(c : Container)
              → over ports (λ(ps : List Natural) → p # [ 22 ]) c
            )
      )
      example
```

... or using the original Kubernetes example, it would be:

```dhall
over
  metadata.labels
  (   λ(xs : Prelude.Map.Type Text Text)
    → xs # toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
  )
  someDeployment
```

It's also easy to implement because you can define the behavior of something like `over`/`set`/`view` inductively, like this:

```dhall
over l.ls… f x = x // { l = over ls… f x.l }

over id f x = f x

set l.ls… e x = x // { l = set ls… e x.l }

set id e x = e

view l.ls… x = view ls… x.l

view id x = x
```

**Comment by sjakobi** [comment]
@Gabriel439 This looks great!

Is there a good reason to add `view` though? It's just the same as field selection, right? On second thought, it might be useful for learning the concept of lenses…

I also wonder whether we should pick a different operator than `.` for composing lenses, to reduce confusion with field selection. OTOH, since field selection and lenses are so related, maybe it's OK when they look similar…

**Comment by Gabriel439** [comment]
@sjakobi: I only included `view` for completeness, but we don't have to standardize that

**Comment by Gabriel439** [comment]
After further reflection, I think there is an even simpler approach to solving this problem that will also solve two other user requests:

1. Reopen https://github.com/dhall-lang/dhall-lang/issues/220 and implement it

   This is the issue proposing Nix-like syntax for nested records (e.g. `{ x.y = 1 }` as short-hand for `{ x = { y = 1 } }`)

   However, the Nix-like syntax would *not* be syntactic sugar for the longer form, because of the following additional change:

2. Reopen https://github.com/dhall-lang/dhall-lang/issues/114 and fix it by change `//` to support deep-overrides using Nix-like syntax

   ```dhall
   r // { x.y = 1 } = r // { x = r.x // { y = 1 } }

   -- NOT the same as:
   r // { x = { y = 1 } }
   ```

   In other words, the compact Nix syntax would behave differently from the long-form syntax in one context: the record override operator.

Now let me explain how I arrived at that particular solution.

The problem I was trying to solve was how to ergonomically support multiple overrides.  I wasn't pleased with the idea of users having to write:

```dhall
set a.b x (set c.d.e y someRecord)
```

... in order to be able to set multiple fields.  My intuition was that this syntax would turn people off due to:

* Excessive expression nesting and parentheses
* Users preferring multiple overrides to be horizontally aligned

So I thought: why not have a syntax like:

```dhall
set { a.b = x, c.d.e = y } someRecord
```

Then I realized that if I standardized the Nix-like syntax for records, then `set` was almost exactly the same as the existing `//` operator.  If you add in the tweak to have the `//` operator give special treatment to the Nix-like syntax then you don't even need to add a new keyword, plus you fix #114 at the same time.

That only leaves one last thing: how to support applying a function to a nested field.  The conclusion I came to was that at least in the short term we could just get + set to modify a field, like this:

```dhall
x // { a.b.c = f x.a.b.c }
```

... which I figured most users would be okay with.  If I'm wrong and if users really object to that idiom then we can standardize an operator or keyword for that separately as a follow-up proposal.

The two running examples would look like this under the Nix-like approach:

```dhall
let Container = { ports : List Natural }

in    example
    ⫽ { spec.template.spec.containers =
          Prelude.List.map
          Container
          Container
          (λ(c : Container) → c // { ports = c.ports # [ 22 ] })
          example.spec.template.spec.containers
      }
```

```dhall
  someDeployment
⫽ { metadata.labels =
        someDeployment.metadata.labels
      # toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
  }
```

The other nice thing about this approach is that it fixes #114 without being a breaking change to how `//` works.

**Comment by timbertson** [comment]
That sounds promising @Gabriel439, although I'm worried that the nixy-ness of a record will become a confusing source of difference which isn't reflected in the types. e.g.:

```
someRecord // { x.y.z = 1 }
```

That's clearly a nested override. But what about:

```
let override = { x.y.z = 1 } in
someRecord // override
```

is that still a nested override. And how about when `override` is:

 - `{ x.y.z = 1 } : { x : { y : { z : Natural } } }` (does the explicit type mean this stops being an override for x.y.z?)
 - `{ x.y = { z = 1 } }` (overrides x.y?)
 - `{ x.y.z = 1 } // { a.b.c = 1 }` (overrides both x.y.z and a.b.c? Or is the nixy-ness lost by the `//` operator and the resulting record is treated as if it were a plain `{ x = { ... }, a = { ... } }` )

Thinking it through, it seems like it'd have to be a syntax thing, so you'd only get the new behaviour if the RHS is a literal. If so, I'd prefer to see:

 - standardise the nix syntax, but
 - create a new operator for this operation. I like your `set { a.b = x, c.d.e = y } someRecord` although I'd prefer to have the override second, so maybe `someRecord with { a.b = x, c.d.e = y }`. And to avoid ambiguity, the RHS _must_ be a literal for this operator.

**Comment by Gabriel439** [comment]
@timbertson: Yeah, that seems reasonable to me.  I also should add that this would also imply changing the `::` record completion operator to use `with` instead of `//` so that it would support deep overriding of defaults

**Comment by sjakobi** [comment]
I like the idea of having

```dhall
r with { a.b = x, c = y }
```

desugar to

```dhall
r // { a = r.a // { b = x }, c = y }
```

That would nicely address the nested-override problem.

The need for short-hand record literals (`{ x.y = 1 }` as short-hand for `{ x = { y = 1 } }`) is less obvious to me, and I think its proposed syntax would cause confusion with the syntax for nested overrides.

Although maybe this could be resolved by choosing a syntax for nested overrides that emphasizes that the thing on the RHS is not a regular record…

**Comment by sjakobi** [comment]
If we introduce this construct of "lens paths" for nested overrides, we could possibly reuse it later for something akin to `lens`'s `over`, where e.g.

    r withf { a.b = f, c = g }

would mean

    r // { a = r.a // { b = f r.a.b }, c = g r.c }

or, more compactly,

    r with { a.b = f r.a.b, c = g r.c }

`withf` is pretty terrible though…I hope we can find a better syntax. :laughing:

**Comment by Gabriel439** [comment]
@sjakobi: Yeah, you are correct that we could standardize support for deep overrides even before standardizing the Nix-like record syntax (taking care to treat the override as syntactically constrained literal rather than an arbitrary Dhall expression like @timbertson suggested)

Regarding `withf`, I think in practice that doing `r withf { a.b = f }` will in the common case be less ergonomic than `r with { a.b = f a.b }` in most cases.  For example, consider this expression:

```dhall
someDeployment with {
, metadata.labels =
      someDeployment.metadata.labels
    # toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
}
```

If you were to define that using `withf` you'd have to annotate the type of the previous value for the `metadata.labels` field:

```dhall
someDeployment withf {
, metadata.labels = λ(labels : Prelude.Map.Type Text Text) →
      labels
    # toMap { `scheduler.alpha.kubernetes.io/critical-pod` = "" }
}
```

**Comment by Gabriel439** [comment]
Just an update: I'm planning to standardize this in three steps:

1. Standardize support for desugaring repeated occurrences of a field to `∧`

   In other words, `{ foo = x, foo = y }` is now legal and is syntactic sugar for `{ foo = x ∧ y }`

   This will pave the way for a very simple implementation of `{ foo.bar.baz = x }`

2. Standardize support for Nix-like record syntax

   This will treat `{ foo.bar.baz = x }` as syntactic sugar for `{ foo = { bar = { baz = x } } }`

   Since this is syntactic sugar it would entail no other changes to the standard, meaning that type-checking, normalization, and the binary encoding are not aware of this change since they only work with the expanded form of the record.  This would also simplify implementing this change since they don't need to change how they internally implement record literals.

   This pairs well with the previous change because if you do:

   ```dhall
   { a.b.c = x, a.b.d = y }
   ```

   ... then that will desugar to:

   ```dhall
   { a = { b = { c = x } }, a = { b = { d = y } } }
   ```

   ... which will in turn desugar to:

   ```dhall
   { a = { b = { c = x } } ∧ { b = { d = y } } }
   ```

3. Standardize support for `with`

   `with` would just work with ordinary record literals (like @sjakobi suggested) and the Nix-like syntax would just be syntactic sugar that would be desugared first before the rules for `with` take effect

**Comment by sjakobi** [comment]
> 1. Standardize support for desugaring repeated occurrences of a field to `∧`
>
>    In other words, `{ foo = x, foo = y }` is now legal and is syntactic sugar for `{ foo = x ∧ y }`

That makes sense in the context of the other planned changes, but is somewhat surprising behaviour on its own IMHO. I'd even imagine accidental bugs like

```dhall
{ a = x
, ... -- lots of other fields
, b = y -- Oops, I overlooked that field
, ...
, b = z
}
```

So maybe we ought to have a warning when the very same "field path" is used repeatedly:

``` 
{ a = x, a = y }

=> warning
```

```
{ a.b = x, a.c = y }

=> Ok!
```

Or `dhall lint` could rewrite

```
{ a = x, a = y }
```

to 

```
{ a = x /\ y }
```




> 3\. Standardize support for `with`
>     `with` would just work with ordinary record literals (like @sjakobi suggested) and the Nix-like syntax would just be syntactic sugar that would be desugared first before the rules for `with` take effect

Could you summarize how `with` would work with plain/desugared records? The old description using the "field path" syntax made sense to me, but I would like to see how it would translate to ordinary records.

**Comment by Profpatsch** [comment]
>     1. `{ foo = x, foo = y }` is now legal and is syntactic sugar for `{ foo = x ∧ y }`

whoa, are you sure this is a good idea? Usually it’s an oversight if two fields are named the same.

Also what happens with `{ foo = { bar = 42; }; } // { foo = { baz = 23 } }`?

This is not at all an obvious improvement to me.

---

What’s the override semantics of `/\` again? It’s not noted in the documentation:
![screenshot](https://user-images.githubusercontent.com/3153638/73011477-977c5f80-3e14-11ea-9133-ee9458c5334e.png)

Will `{ foo = { a = 42 }, foo = { a = 43 } }` be valid?

What would the error message of `{ foo = { a = 42 }, foo = 43 }` be?

**Comment by ocharles** [comment]
Another very raised eyebrow to allowing multiple occurrences of a field. This feels very subtle, and in a large Dhall document you could end up thinking you were adding a field, when in reality you were appending to a field (e.g., for large records that do not fit on one screen).

**Comment by Gabriel439** [comment]
@sjakobi @Profpatsch: All the examples you both were concerned about would be type errors:

```haskell
{ foo = 42, foo = 43 }  -- Type error: 42 ∧ 43 is ill-typed

{ foo = { a = 42 }, foo = 43 }  -- Type error: { a = 42 } ∧ 43 is ill-typed

{ foo = { a = 42 }, foo = { a = 43 } }  -- Type error: { a = 42 } ∧ { a = 43 } is ill-typed
```

**Comment by Gabriel439** [comment]
@sjakobi: `with` would only work with normal desugared records (and if you used the Nix-like record syntax for the override would be desugared before evaluating `with`).  The behavior is actually very straightforward to implement and standardize:

```dhall
-- Just reiterating that Nix-like syntax is desugared first
r with { k₀.k₁ = v, … } = r with { k₀ = { k₁ = v, … } }

-- Inductive case
r with { k₀ = v₀, k₁ = v₁, … } = r // { k₀ = r.k₀ with v₀, k₁ = r.k₁ with v₁, … }

-- Base case, when `v` is not a record literal
r with v = v
```

**Comment by philandstuff** [comment]
Apologies, this is a bit of a drive-by review, but:

One case I’m still a bit confused/concerned about: what would `{ a = foo, a.b = bar }` mean?

Basically +1 to “this feels subtle and confusing and like it has traps for the unwary”. I wonder if there is a way of trying it out before adding it to the standard? I’d like to be able to give it a go, but in a way that we can back out of it if the experiment doesn’t work for some reason.

**Comment by Gabriel439** [comment]
@philandstuff:

```dhall
{ a = foo, a.b = bar }

= { a = foo, a = { b = bar } }

= { a = foo ∧ { b = bar } }
```

... and whether or not that is well-typed depends on the type of `foo` and it follows the normal type-checking rules for `∧`

**Comment by Gabriel439** [comment]
Here is the pull request for the first change (desugaring duplicate record fields): https://github.com/dhall-lang/dhall-lang/pull/896

The pull request description and test files provide more detail about why I chose to factor things in this way

**Comment by Gabriel439** [comment]
The second pull request for dotted field syntax is up here: #901

**Comment by Gabriel439** [comment]
The third pull request for the `with` keyword is up here: https://github.com/dhall-lang/dhall-lang/pull/923

**Comment by TristanCacqueray** [comment]
@Gabriel439 that issue doesn't seem to be solved by the current `with` implementation:

```
$ dhall <<< '{ foo = [41] } with foo = [42]'
{ foo = [ 42 ] }
```

Did I missed something or should this be re-opened?

**Comment by sjakobi** [comment]
@TristanCacqueray I think that original idea was abandoned somewhere in this thread.

Appending to a list in a nested record field looks somewhat like this now:

```dhall
someDeployment with metadata.labels =
  someDeployment.metadata.labels # additionalLabels
```

**Comment by TristanCacqueray** [comment]
@sjakobi alright, thank you for the summary :)
